scalafix-core/shared/src/main/scala/scalafix/internal/rule/DiscardedValues.scala

```diff
@@ -0,0 +1,80 @@
+package scalafix.internal.rule
+
+import scala.meta.Term.Block
+import scala.meta._
+import scalafix.SemanticdbIndex
+import scalafix.internal.util.TypeSyntax
+import scalafix.lint.{LintCategory, LintMessage}
+import scalafix.rule.{RuleCtx, RuleName, SemanticRule}
+import scalafix.syntax._
+import scala.meta.contrib.implicits.Equality._
+
+case class DiscardedValues(index: SemanticdbIndex)
+    extends SemanticRule(index, RuleName("DiscardedValues")) {
+
+  override def description: String =
+    "This rule ensures that statements should only return Unit"
+
+  private lazy val errorCategory: LintCategory =
+    LintCategory.error("Statements should only return Unit")
+
+  def litType(l: Lit): Option[Type] = l match {
+    case Lit.Int(_) => Some(t"_root_.scala.Int")
+    case Lit.Unit() => Some(t"_root_.scala.Unit")
+    case Lit.Float(_) => Some(t"_root_.scala.Float")
+    case Lit.Double(_) => Some(t"_root_.scala.Double")
+    case Lit.Symbol(_) => Some(t"_root_.scala.Symbol")
+    case Lit.Boolean(_) => Some(t"_root_.scala.Boolean")
+    case Lit.Byte(_) => Some(t"_root_.scala.Byte")
+    case Lit.Char(_) => Some(t"_root_.scala.Char")
+    case Lit.Long(_) => Some(t"_root_.scala.Long")
+    case Lit.Null() => Some(t"_root_.scala.Null")
+    case Lit.String(_) => Some(t"_root_.scala.String")
+    case Lit.Short(_) => Some(t"_root_.scala.Short")
+    case _ => None
+  }
+
+  def statementType(stat: Stat): Option[Type] = stat match {
```

Comment:
This is an incomplete solution, there is no way to robustly get the type of a term with the current scalafix API. If there was supported there would be a method for it in the public API 😜  See "Type at position => No" in http://scalameta.org/

You are pretty-much writing your own type-checker here! It's not a bad approach and it will work for some cases, but I think we should do this the Right Way ™ and use the compiler for this kind of stuff.